File: demo/src/main/java/design/pattern/lg/DipDemo.java
package design.pattern.lg;

import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;

class StringProcessor { //具体类
    private final StringReader stringReader; //具体类
    private final StringWriter stringWriter; //具体类

    //通过构造函数来注入依赖组件
    public StringProcessor(StringReader stringReader, StringWriter stringWriter) {
        this.stringReader = stringReader;
        this.stringWriter = stringWriter;
    }

    public void readAndWrite() throws IOException {
        stringWriter.write(stringReader.read());
    }

    //测试用例

    public static void main(String[] args) throws IOException {
        StringReader sr = new StringReader("1111111");
        sr.read();
        StringWriter sw = new StringWriter();
        StringProcessor sp = new StringProcessor(sr, sw);
        sp.readAndWrite();
    }
}

interface IStringProcessor {
    void readAndWrite(StringReader stringReader, StringWriter stringWriter) throws IOException;
}

class StringProcessorImpl implements IStringProcessor {

    //通过接口传参来注入依赖组件
    @Override
    public void readAndWrite(StringReader stringReader, StringWriter stringWriter) throws IOException {
        stringWriter.write(stringReader.read());
    }

    public static void main(String[] args) throws IOException {
        StringReader sr = new StringReader("222222");
        sr.read();
        StringWriter sw = new StringWriter();
        IStringProcessor sp = new StringProcessorImpl();
        sp.readAndWrite(sr, sw);
    }
}

interface IStringReader {
    void read(String path);

    String getValue();
}

interface IStringWriter {
    void write(String value);
}

interface IStringProcessor2 {
    void readAndWrite(IStringReader stringReader, IStringWriter stringWriter) throws IOException;
}

class StringReaderImpl implements IStringReader {
    @Override
    public void read(String path) {
    }

    @Override
    public String getValue() {
        return null;
    }
}

class StringWriterImpl implements IStringWriter {
    @Override
    public void write(String value) {
    }
}

class StringProcessorImpl2 implements IStringProcessor2 {
    @Override
    public void readAndWrite(IStringReader stringReader, IStringWriter stringWriter) {
        stringWriter.write(stringReader.getValue());
    }

    public static void main(String[] args) throws IOException {
        IStringReader sr = new StringReaderImpl();
        sr.read("333333");
        IStringWriter sw = new StringWriterImpl();
        IStringProcessor2 sp = new StringProcessorImpl2();
        sp.readAndWrite(sr, sw);
    }
}

class SPTest {
    //@Resource
    private StringProcessor sp;

    //@Resource
    private StringReader sr;

    //@Resource
    private StringWriter sw;

    public void main(String[] args) {
        // sr.read("444444");
        // sp.readAndWrite(sr, sw);
    }
}

File: demo/src/main/java/design/pattern/lg/DipDemo.kt
package design.pattern.lg

import java.io.IOException
import java.io.StringReader
import java.io.StringWriter

internal class StringProcessor     //通过构造函数来注入依赖组件
    (//具体类
    private val stringReader //具体类
    : StringReader, //具体类
    private val stringWriter: StringWriter
) {
    @Throws(IOException::class)
    fun readAndWrite() {
        stringWriter.write(stringReader.read())
    }

    companion object {
        //测试用例
        @Throws(IOException::class)
        @JvmStatic
        fun main(args: Array<String>) {
            val sr = StringReader("1111111")
            sr.read()
            val sw = StringWriter()
            val sp = StringProcessor(sr, sw)
            sp.readAndWrite()
        }
    }
}

internal interface IStringProcessor {
    @Throws(IOException::class)
    fun readAndWrite(stringReader: StringReader, stringWriter: StringWriter)
}

internal class StringProcessorImpl : IStringProcessor {
    //通过接口传参来注入依赖组件
    @Throws(IOException::class)
    override fun readAndWrite(stringReader: StringReader, stringWriter: StringWriter) {
        stringWriter.write(stringReader.read())
    }

    companion object {
        @Throws(IOException::class)
        @JvmStatic
        fun main(args: Array<String>) {
            val sr = StringReader("222222")
            sr.read()
            val sw = StringWriter()
            val sp: IStringProcessor = StringProcessorImpl()
            sp.readAndWrite(sr, sw)
        }
    }
}

internal interface IStringReader {
    fun read(path: String?)
    val value: String?
}

internal interface IStringWriter {
    fun write(value: String?)
}

internal interface IStringProcessor2 {
    @Throws(IOException::class)
    fun readAndWrite(stringReader: IStringReader, stringWriter: IStringWriter)
}

internal class StringReaderImpl : IStringReader {
    override fun read(path: String?) {}
    override val value: String?
        get() = null
}

internal class StringWriterImpl : IStringWriter {
    override fun write(value: String?) {}
}

internal class StringProcessorImpl2 : IStringProcessor2 {
    override fun readAndWrite(stringReader: IStringReader, stringWriter: IStringWriter) {
        stringWriter.write(stringReader.value)
    }

    companion object {
        @Throws(IOException::class)
        @JvmStatic
        fun main(args: Array<String>) {
            val sr: IStringReader = StringReaderImpl()
            sr.read("333333")
            val sw: IStringWriter = StringWriterImpl()
            val sp: IStringProcessor2 = StringProcessorImpl2()
            sp.readAndWrite(sr, sw)
        }
    }
}

internal class SPTest {
    //@Resource
    private val sp: StringProcessor? = null

    //@Resource
    private val sr: StringReader? = null

    //@Resource
    private val sw: StringWriter? = null
    fun main(args: Array<String?>?) {
        // sr.read("444444");
        // sp.readAndWrite(sr, sw);
    }
}